datalad_next/patches/replace_sshremoteio.py

```diff
@@ -0,0 +1,397 @@
+        # lean, even for big archives
+        cmd = f'7z l {sh_quote(str(archive_path))} {sh_quote(loc)}'
+
+        # Note: Currently relies on file_path not showing up in case of failure
+        # including non-existent archive. If need be could be more sophisticated
+        # and called with check=True + catch RemoteCommandFailedError
+        out = self.servershell(
+            cmd,
+            check=False,
+        ).stdout.decode()
+
+        return loc in out
+
+    def get_from_archive(self, archive, src, dst, progress_cb):
+
+        # Note, that as we are in blocking mode, we can't easily fail on the
+        # actual get (that is 'cat'). Therefore check beforehand.
+        if not self.exists(archive):
+            raise RIARemoteError("archive {arc} does not exist."
+                                 "".format(arc=archive))
+
+        # TODO: We probably need to check exitcode on stderr (via marker). If
+        #       archive or content is missing we will otherwise hang forever
+        #       waiting for stdout to fill `size`.
+
+        cmd = '7z x -so {} {}\n'.format(
+            sh_quote(str(archive)),
+            sh_quote(str(src)))
+        self.shell.stdin.write(cmd.encode())
+        self.shell.stdin.flush()
+
+        # TODO: - size needs double-check and some robustness
+        #       - can we assume src to be a posixpath?
+        #       - RF: Apart from the executed command this should be pretty much
+        #         identical to self.get(), so move that code into a common
+        #         function
+
+        from os.path import basename
+        size = self._get_download_size_from_key(basename(str(src)))
+
+        with open(dst, 'wb') as target_file:
+            bytes_received = 0
+            while bytes_received < size:
+                c = self.shell.stdout.read1(self.buffer_size)
+                if c:
+                    bytes_received += len(c)
+                    target_file.write(c)
+                    progress_cb(bytes_received)
+
+    def read_file(self, file_path):
+
+        cmd = f"cat {sh_quote(str(file_path))}"
+        try:
+            out = self.servershell(
+                cmd,
+                check=True,
+            ).stdout.decode()
+        except CommandError as e:
+            # Currently we don't read stderr. All we know is, we couldn't read.
+            # Try narrowing it down by calling a subsequent exists()
+            if not self.exists(file_path):
+                raise FileNotFoundError(f"{str(file_path)} not found.") from e
+            else:
+                raise RuntimeError(f"Could not read {file_path}") from e
+
+        return out
+
+    def write_file(self, file_path, content, mode='w'):
+
+        if mode == 'w':
+            mode = ">"
+        elif mode == 'a':
+            mode = ">>"
+        else:
+            raise ValueError("Unknown mode '{}'".format(mode))
+        if not content.endswith('\n'):
+            content += '\n'
```

Comment:
@christian-monch If I remove this safeguard and the trailing newline from the corresponding test, things stop working. What is necessary to not have to require a trailing newline?